src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,183 @@
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = getNode(index).item;
+        getNode(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void linkFirst(T t) {
+        final Node<T> head = this.head;
+        final Node<T> newNode = new Node<>(null, t, head);
+        this.head = newNode;
+        if (head == null) {
+            this.tail = newNode;
+        } else {
+            head.prev = newNode;
+        }
+        size++;
+    }
+
+    void linkLast(T t) {
+        final Node<T> tail = this.tail;
+        final Node<T> newNode = new Node<>(tail, t, null);
+        this.tail = newNode;
+        if (tail == null) {
+            this.head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        size++;
+    }
+
+    void linkBefore(T t, Node<T> current) {
+        final Node<T> pred = current.prev;
+        final Node<T> newNode = new Node<>(pred, t, current);
+        current.prev = newNode;
+        if (pred == null) {
+            head = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(getErrorMessage(index));
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> x = null;
+        if (index < (size >> 1)) {
```

Comment:
This better for readability
```suggestion
        if (index < (size / 2)) {
```